### PR TITLE
Replacing solve with solveset for solving inequalities

### PR DESCRIPTION
**Pull request by kshitij10496**
This PR is an attempt to use the results from `solveset` for solving univariate inequalities (in the real domain). Currently, the primary method for solving such inequalities - `solve_univariate_inequality`, uses the `solve` internally. 
- [x] implement `solvify`
- [x] XFAIL failing tests
- [x] docstrings
- [x] consolidate `solve_univariate_inequality` function for the new `domain` argument.
- [x] added tests from #9721

### CODE REVIEW

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -811,6 +811,46 @@ def solveset_complex(f, symbol):
     return solveset(f, symbol, S.Complexes)
 
 
+def solvify(f, symbol, domain=S.Reals):
```
Comment (hargup) [inline sympy/solvers/solveset.py]: Nice name, `solvify` the output of `solveset`.

Comment (hargup) [inline sympy/solvers/solveset.py]: I would prefer to be function which takes a set and returns the solution in a form which `solve` would return. So the usage would be `solvify(solveset(f, symbol, domain))`

Comment (kshitij10496) [inline sympy/solvers/solveset.py]: > I would prefer to be function which takes a set and returns the solution in a form which solve would return. So the usage would be solvify(solveset(f, symbol, domain))

This sounds cool as well.

Comment (kshitij10496) [inline sympy/solvers/solveset.py]: Giving it a second thought, we need the function `f` to try and compute its period (via `periodicity`).
This function is true to its cause even if we use `solvify(f, symbol, domain)`(by internalising the call to `solveset`). 
So, I think that using the current implementation is a good idea.

**Comment by kshitij10496** [comment]
Ping @sympy/sympy-solvers 
This PR is ready for review once Travis builds successfully.

**Comment by kshitij10496** [comment]
This PR fixes issue #9721 as well.

From master
```python
In []: solve_univariate_inequality(cos(x) > S(0), x, relational=False)
Out[]: (-oo, pi/2)

In []: solve_univariate_inequality(tan(x) > S(0), x, relational=False)
Out[]: (0, oo)


```
From this branch
```python
In []: from sympy.solvers.inequalities import solve_univariate_inequality as isolve

In []: isolve(cos(x) > S(0), x, relational=False)
Out[]: (0, pi/2) U (3*pi/2, 2*pi)

In []: isolve(tan(x) > S(0), x, relational=False)
Out[]: (0, pi/2)

```

**Comment by kshitij10496** [comment]
Ping @sympy/sympy-solvers

File: sympy/core/tests/test_wester.py [hunk 2]
```diff
@@ -1223,7 +1223,7 @@ def test_N12():
 
 def test_N13():
     x = Symbol('x')
-    assert solveset(sin(x) < 2, domain=S.Reals) == S.Reals
+    assert solveset(sin(x) < 2, domain=S.Reals) == Interval(0, 2*pi, False, True)
```
Comment (Shekharrajak) [inline sympy/core/tests/test_wester.py]: Ans should be all +ve real numbers, isn't it ?

File: sympy/solvers/tests/test_solveset.py [hunk 3]
```diff
@@ -932,6 +935,19 @@ def test_issue_9522():
...
+def test_solvify():
+    x = Symbol('x')
+
+    assert solvify(x**2 + 10, x, S.Reals) == []
+    assert solvify(x**3 + 1, x, S.Complexes) == [-1, 1/2 - sqrt(3)*I/2,
+                                                 1/2 + sqrt(3)*I/2]
+    assert solvify(log(x), x, S.Reals) == [1]
+    assert solvify(cos(x), x, S.Reals) == [pi/2, 3*pi/2]
+    assert solvify(sin(x) + 1, x, S.Reals) == [3*pi/2]
+    assert solvify(sin(Abs(x)), x, S.Reals) is None
```
Comment (Shekharrajak) [inline sympy/solvers/tests/test_solveset.py]: Ans is `0, pi` ?

File: sympy/core/tests/test_wester.py (hunk 2 above)
Comment (kshitij10496) [inline sympy/core/tests/test_wester.py]: > Ans should be all +ve real numbers, isn't it ?

Yes. You are mathematically correct.
I have stated the reason for doing so in my blog [post](http://kshitij10496.github.io/2016/08/07/GSoC-Week-10-11/).

Restricting the domain of solution now, will help us later in extension (with the implementation of `BigUnion`).

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_solveset.py]: > Ans is 0, pi ?

```python
In []: solveset(sin(Abs(x)), x, S.Reals)
Out[]: Intersection((-oo, 0], ImageSet(Lambda(_n, _n*pi), Integers())) U Intersection([0, oo), ImageSet(Lambda(_n, _n*pi), Integers()))
```
The solution returned by `solveset` is independent of the assumption on `x`.
Hence, an Intersection of intervals with imageset is returned.
And, currently, `solvify` doesn't handle the case of `Intersection` solutions.

**Comment by kshitij10496** [comment]
Ping @sympy/sympy-solvers

File: sympy/solvers/inequalities.py [hunk 4]
```diff
@@ -378,7 +379,7 @@ def reduce_abs_inequalities(exprs, gen):
         for expr, rel in exprs ])
 
 
-def solve_univariate_inequality(expr, gen, relational=True):
+def solve_univariate_inequality(expr, gen, domain=S.Reals, relational=True):
```
Comment (moorepants) [inline sympy/solvers/inequalities.py]: This new kwarg should be added after relational to preserve the order and backwards compatibility.

File: sympy/solvers/tests/test_inequalities.py [hunk 5]
```diff
@@ -323,15 +337,16 @@ def test_issue_10198():
...
+        And(-oo < x, x < 0), Eq(x, 1))
     assert reduce_abs_inequality(-3 + 1/abs(1 - 1/x), '<', x) == \
         Or(And(-oo < x, x < 0),
         And(S(0) < x, x < S(3)/4), And(S(3)/2 < x, x < oo))
     raises(ValueError,lambda: reduce_abs_inequality(-3 + 1/abs(
         1 - 1/sqrt(x)), '<', x))
 
 def test_issue_10047():
-    assert solve(sin(x) < 2) == And(-oo < x, x < oo)
+    assert solve(sin(x) < 2) == And(S(0) <= x, x < 2*pi)
```
Comment (moorepants) [inline sympy/solvers/tests/test_inequalities.py]: It's generally not a good idea to change unit tests, especially when these are here for regression purposes for a specific solved issue.

File: sympy/solvers/inequalities.py (hunk 4 above)
Comment (kshitij10496) [inline sympy/solvers/inequalities.py]: > This new kwarg should be added after relational to preserve the order and backwards compatibility.

I think I overlooked this fact because everywhere `relational=bool` is used.
Irrespective of this, it is a blunder on my part.
Thanks for pointing this out.
 I will make the changes.

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (Shekharrajak) [inline sympy/solvers/tests/test_solveset.py]: So whenever solveset is returning `Intersection` soln, inequality solvers `solve_univariate_inequality` will not work. Same for `complement` soln ? Can you write about it in docstring ?

File: sympy/stats/tests/test_symbolic_probability.py [hunk 6]
```diff
@@ -9,7 +9,8 @@ def test_literal_probability():
     Y = Normal('Y', 3, 4)
     Z = Poisson('Z', 4)
     W = Poisson('W', 3)
-    x, y, w, z = symbols('x, y, w, z')
+    x = symbols('x', real=True)
+    y, w, z = symbols('y, w, z')
```
Comment (Shekharrajak) [inline sympy/stats/tests/test_symbolic_probability.py]: Why this changes?

File: sympy/solvers/inequalities.py [hunk 7]
```diff
@@ -411,16 +413,24 @@ def solve_univariate_inequality(expr, gen, relational=True):
     elif expr is S.false:
         rv = S.EmptySet
     else:
+        solns = None
```
Comment (Shekharrajak) [inline sympy/solvers/inequalities.py]: No need of this line.

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_solveset.py]: > So whenever solveset is returning Intersection soln, inequality solvers solve_univariate_inequality will not work. Same for complement soln ? 

Yes.

> Can you write about it in docstring ?

I have mentioned this in `solvify` function.
However, it would be a good idea to state this behaviour in `solve_univariate_inequality` as well.
Thanks ! :)

File: sympy/solvers/solveset.py [hunk 8]
```diff
@@ -882,6 +883,86 @@ def solveset_complex(f, symbol):
     return solveset(f, symbol, S.Complexes)
 
 
+def solvify(f, symbol, domain):
+    """Solves an equation and returns the solution in accordance with the
```
Comment (hargup) [inline sympy/solvers/solveset.py]: `Solves an equation using solveset ...`

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (hargup) [inline sympy/solvers/tests/test_solveset.py]: Also add it as an `XFAIL` test.

File: sympy/solvers/solveset.py (hunk 8 above)
Comment (kshitij10496) [inline sympy/solvers/solveset.py]: > should be Solves an equation using solveset ...

Sounds better.

File: sympy/solvers/tests/test_inequalities.py (hunk 5 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: > It's generally not a good idea to change unit tests, especially when these are here for regression purposes for a specific solved issue.

Yes. 
In this scenario, we have changed the solving technique.
We have restricted the domain of the solution to a periodic interval.
I changed the test because I thought that it would be better to display this behaviour throughout the test suite.

Should I XFAIL this or modifying the test is better ?
Ping @moorepants @hargup @aktech

Comment (hargup) [inline sympy/solvers/tests/test_inequalities.py]: I feel that you should change the behavior of `solve` and use this PR only to internally replace solve with `solveset` to places where you can. If you want a different, maybe better, behavior of inequalities do it in `solveset` instead.

Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: The whole test suite is passing so I think that the change in behaviour is backward compatible with `solve`. Since, we have followed (as well as expected) this behaviour throughout our test suite, I personally think that this test case can be modified in accordance with the restricted domain feature.

With regards to `solveset`, I was thinking that we can return the same result for now.
After the implementation of `BigUnion` and `BigIntersection`, we can extend the solution set (returned by `solve_univariate_inequality`) to return the complete solution.

What are your thoughts on this ?

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_solveset.py]: > Also add it as an XFAIL test.

I have added an XFAIL test [here](https://github.com/sympy/sympy/pull/11458/files#diff-b3fe41ada1b7b645fe6a140b3ae5da32R359)

File: sympy/solvers/inequalities.py (hunk 7 above)
Comment (kshitij10496) [inline sympy/solvers/inequalities.py]: I have removed it.
Thanks

File: sympy/core/tests/test_wester.py (hunk 2 above)
Comment (hargup) [inline sympy/core/tests/test_wester.py]: As of now, your result is incorrect and you should not implicitly restrict the domain of the function. When you have ability to represent infinite unions or intersections then you can make the other required changes.

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (hargup) [inline sympy/solvers/tests/test_solveset.py]: > And, currently, `solvify` doesn't handle the case of `Intersection` solutions.

This also reminds us that we need better ways to extract solutions from nested `intersections` and `unions`.

Comment (kshitij10496) [inline sympy/solvers/tests/test_solveset.py]: > This also reminds us that we need better ways to extract solutions from nested intersections and unions.

Yes. We need to.
I think @aktech suggested implementing a wrapper class to do this.

File: sympy/core/tests/test_wester.py (hunk 2 above)
Comment (kshitij10496) [inline sympy/core/tests/test_wester.py]: > As of now, your result is incorrect and you should not implicitly restrict the domain of the function.

I agree this is not the complete answer.

> When you have ability to represent infinite unions or intersections then you can make the other required changes.

Should we XFAIL this for now ?

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (aktech) [inline sympy/solvers/tests/test_solveset.py]: I thought a bit about it, will try to send a PR soon.

Comment (kshitij10496) [inline sympy/solvers/tests/test_solveset.py]: Cool. :+1:

File: sympy/solvers/tests/test_inequalities.py (hunk 5 above)
Comment (Shekharrajak) [inline sympy/solvers/tests/test_inequalities.py]: Why not define `solveset_univariate_inequality` for `solveset`, rather than modifying `solve_univariate_inequality` ?. To get complete solution we should not edit old one, right? For these changes it is fine (since it is returning soln within periodic interval), but for complete solution it will use `ImageSet`, isn't it ?

Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: > Why not define solveset_univariate_inequality for solveset, rather than modifying solve_univariate_inequality ?

The logic of the new function would fundamentally be the same.
Moreover, the current implementation of `solve_univariate_inequality` is robust.

> To get complete solution we should not edit old one, right?

Please elaborate on why do you think so ?

> For these changes it is fine (since it is returning soln within periodic interval), but for complete solution it will use ImageSet, isn't it ?

No. It will use `BigUnion`

File: sympy/stats/tests/test_symbolic_probability.py (hunk 6 above)
Comment (kshitij10496) [inline sympy/stats/tests/test_symbolic_probability.py]: The solver depends upon the assumption of `x` being a real number.

**Comment by kshitij10496** [comment]
I think I have addressed all the concerns.
Let us have another round of review.

Ping @hargup @aktech

File: sympy/solvers/tests/test_inequalities.py [hunk 9]
```diff
@@ -291,6 +286,27 @@ def test_solve_univariate_inequality():
...
+    assert isolve(cos(x) < S.Zero, x, relational=False) == \
+        Interval(pi/2, 3*pi/2, True, True)
+    assert isolve(cos(x) >= S.Zero, x, relational=False) == \
+        Union(Interval(0, pi/2), Interval(3*pi/2, 2*pi, False, True))
+
+    assert isolve(tan(x) < S.One, x, relational=False) == \
+        Union(Interval(0, pi/4, False, True), Interval(pi/2, pi, True, True))
+
+    assert isolve(sin(x) <= S.Zero, x, relational=False) == \
+        Union(FiniteSet(S(0)), Interval(pi, 2*pi, False, True))
```
Comment (Shekharrajak) [inline sympy/solvers/tests/test_inequalities.py]: `Interval(pi, 2*pi, False, False))` . `2*pi` is also a soln.

File: sympy/solvers/tests/test_inequalities.py [hunk 10]
```diff
@@ -291,6 +286,27 @@ def test_solve_univariate_inequality():
...
+def test_trig_inequalities():
+    # all the inequalities are solved in a periodic interval.
+    assert isolve(sin(x) < S.Half, x, relational=False) == \
+        Union(Interval(0, pi/6, False, True), Interval(5*pi/6, 2*pi, True, True))
+    assert isolve(sin(x) > S.Half, x, relational=False) == \
+        Interval(pi/6, 5*pi/6, True, True)
+    assert isolve(cos(x) < S.Zero, x, relational=False) == \
+        Interval(pi/2, 3*pi/2, True, True)
+    assert isolve(cos(x) >= S.Zero, x, relational=False) == \
+        Union(Interval(0, pi/2), Interval(3*pi/2, 2*pi, False, True))
```
Comment (Shekharrajak) [inline sympy/solvers/tests/test_inequalities.py]: `2*pi` is also a soln.

File: sympy/solvers/tests/test_inequalities.py [hunk 11]
```diff
@@ -323,16 +340,24 @@ def test_issue_10198():
         -1 + 1/abs(1/x - 1) < 0) == Or(
         And(-oo < x, x < 0), And(S(0) < x, x < S(1)/2)
         )
-    raises(NotImplementedError, lambda: reduce_inequalities(abs(1/sqrt(x)) - 1, x))
+    assert reduce_inequalities(abs(1/sqrt(x)) - 1, x) == Or(
+        And(-oo < x, x < 0), Eq(x, 1))
```
Comment (Shekharrajak) [inline sympy/solvers/tests/test_inequalities.py]: Ans is `+1, -1 `, right ? `abs(1/sqrt(-1))` is one.

**Comment by Shekharrajak** [comment]
```
In [ ]: isolve(sqrt(x) <= 1 ,x,S.Reals)
Out[ ]: x ≤ 1 ∧ -∞ < x
```
Ans is `Interval(0, 1)`.

File: sympy/solvers/tests/test_inequalities.py (hunk 9 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: The inequalities are solved in `[0, 2*pi)`
Since `f(x + p) = f(x)`, it didn't make sense to me to include same repetitive values.
Hence, I have used an open interval.

File: sympy/solvers/tests/test_inequalities.py (hunk 10 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: Addressed above.

File: sympy/solvers/tests/test_inequalities.py (hunk 11 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: > Ans is +1, -1, right ?

No. `-1` can't be an answer.

> `abs(1/sqrt(-1))` is one.
Since, `1/sqrt(-1)` is a complex number( iota `I` ).
Taking its `Abs` returns its magnitude i.e. 1.

We solve inequalities in the real domain only.

**Comment by kshitij10496** [comment]
> In [ ]: isolve(sqrt(x) <= 1 ,x, domain=S.Reals)  
  Out[ ]: x ≤ 1 ∧ -∞ < x  
  Ans is Interval(0, 1).

Yes. Currently, we do not check the inequalities for their domain.
Same goes for inequalities involving `log`. 

```python
In []: solveset(log(x) < 0, x, S.Reals)  

Out[]: (-oo, 1)
```

I think this can be improved upon later.

**Comment by kshitij10496** [comment]
I have added a check so as to solve the inequalities in their real, continuous domains.

For example, 
``` python
In []: isolve(sqrt(x) <= 4, x, relational=False)
Out[]: [0, 2]
```

However, there is a particular case which `solveset` is unable to solve
```python
In []: df = (-2*x + 3)/(x**2 - 3*x)**2

In []: solveset(df, x, Interval.open(1.5, 3))
Out[]: ConditionSet(x, Eq((-2*x + 3)/(x**2*(x - 3)**2), 0), (1.5, 3))

In []: solve(df, x)
Out[]: [3/2]
```

Should we add it as an XFAIL for now ?


```

File: sympy/solvers/tests/test_inequalities.py [hunk 12]
```diff
@@ -323,16 +340,24 @@ def test_issue_10198():
...
     assert reduce_abs_inequality(-3 + 1/abs(1 - 1/x), '<', x) == \
         Or(And(-oo < x, x < 0),
         And(S(0) < x, x < S(3)/4), And(S(3)/2 < x, x < oo))
     raises(ValueError,lambda: reduce_abs_inequality(-3 + 1/abs(
         1 - 1/sqrt(x)), '<', x))
 
+
 def test_issue_10047():
-    assert solve(sin(x) < 2) == And(-oo < x, x < oo)
+    assert solve(sin(x) < 2) == And(S(0) <= x, x < 2*pi)
```
Comment (hargup) [inline sympy/solvers/tests/test_inequalities.py]: My issue is that this solution is wrong, unless you explicitly specify that the domain is `[0, 2*pi)`.

**Comment by hargup** [comment]
> However, there is a particular case which `solveset` is unable to solve
> ```python
> In []: df = (-2*x + 3)/(x**2 - 3*x)**2
> 
> In []: solveset(df, x, Interval.open(S(3)/2, 3))
> Out[]: ConditionSet(x, Eq((-2*x + 3)/(x**2*(x - 3)**2), 0), (1.5, 3))
> ```


This is strange, shouldn't `solveset` simply take an intersection of the domain
with the solution it gets with domain as `S.Reals`?

Please report an issue about that. If it cannot be fixed with some minor
changes leave it for now and do `XFAIL`.

**Comment by hargup** [comment]
Except [this](https://github.com/sympy/sympy/pull/11458#discussion_r74812631) it is plus one after tests passes.

**Comment by kshitij10496** [comment]
>  If it cannot be fixed with some minor
changes leave it for now and do XFAIL.

I think I have fixed the issue.
Apparently, we were not checking if the `inverter` for returning an `EmptySet`.
In this case, we can assuredly state that **no** solutions exist.

File: sympy/solvers/tests/test_inequalities.py (hunk 12 above)
Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: > My issue is that this solution is wrong, unless you explicitly specify that the domain is [0, 2*pi).

Yes. I agree with you that should not be the solution unless the user specifies the domain to be
`[0, 2*pi)`. 
I think it will be better if we do this an XFAIL for now.
With the implementation of `BigUnion`, we would be able to un-XFAIL it later.

I have, however, documented this behaviour [here](https://github.com/sympy/sympy/pull/11458/files#diff-8c6b1223d1a80f79ca2509e14ebae02eR408). 
Shouldn't this suffice for now ?

Comment (hargup) [inline sympy/solvers/tests/test_inequalities.py]: For this particular answer we don't need Infinite union. 

>  I think it will be better if we do this an XFAIL for now.

Why do we need to return a wrong result in the first place? This is a result which was correct earlier and will be wrong after this PR, this is a digression, isn't it?

Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: > To represent this particular answer we don't need Infinite union.

For solving trigonometric inequalities, we can return an infinite union.
In this particular case, the above-said union simplifies to `S.Reals`.
 
> Why do we need to return a wrong result in the first place?

I agree that this looks to be a regression. 
But in the long run, I don't think XFAILing this (for now) will cause us much harm.

Comment (kshitij10496) [inline sympy/solvers/tests/test_inequalities.py]: Ping @hargup

Comment (hargup) [inline sympy/solvers/tests/test_inequalities.py]: > I agree that this looks to be a regression.
> But in the long run, I don't think XFAILing this (for now) will cause us much harm.

This still doesn't explain why we going this path anyway.

**Comment by kshitij10496** [comment]
I have fixed the regression.
Now,

```python
In []: isolve(sin(x) < 2, x, False)
Out []: S.Reals 
```

Ping @hargup

**Comment by kshitij10496** [comment]
Ping @hargup 

Is this PR good to go now ?

**Comment by kshitij10496** [comment]
I think the build is failing due to some issue with Travis.
So, can we restart the build ?

**Comment by hargup** [comment]
> I think the build is failing due to some issue with Travis.
> So, can we restart the build ?

Someone did it before me. The PR looks good to me overall. What is the test coverage?

**Comment by hargup** [comment]
Okay most of the lines you added are covered. Merging.

**Comment by hargup** [comment]
@kshitij10496 Thank you for your work. Thanks! @Shekharrajak @moorepants for reviews.

**Comment by kshitij10496** [comment]
@hargup Thanks a lot for merging this before the deadline. :smile: 

Thank you, everyone, for your reviews.
